src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+
+        """
+        op = SecondaryOperator(
+            name=self.name,
+            domains=self.domains,  # type:ignore[arg-type]
+            children=args,  # type:ignore[arg-type]
+        )
+        op.fetch_data = self.fetch_data
+        return op
+
+    def previous_timestep(self, steps: int = 1) -> SecondaryOperator:
+        """Secondary operators have children which also need to be obtained at
+        the previous time step."""
+
+        op = super().previous_timestep(steps=steps)
+        op.children = [child.previous_timestep(steps=steps) for child in self.children]
+        return op
+
+    def previous_iteration(self, steps: int = 1) -> SecondaryOperator:
+        """Secondary operators have children which also need to be obtained at
+        the previous iteration."""
+        op = super().previous_iteration(steps=steps)
+        op.children = [child.previous_iteration(steps=steps) for child in self.children]
+        return op
+
+    def func(self, *args: FloatType) -> float | np.ndarray | AdArray:
+        """See :meth:`~porepy.numerics.ad.operator_functions.AbstractFunction.func`.
+
+        Note:
+            This class cannot inherit from
+            :class:`~porepy.numerics.ad.operator_functions.AbstractFunction` because
+            that parent would block arithmetic overloads.
+            But it uses the operation ``evaluate`` and the ``AbstractFunction.func``
+            with ``self`` as an explicit argument.
+
+        """
+        # mypy complains that self is not an instance of AbstractFunction because of the
+        # hack
+        return pp.ad.AbstractFunction.func(self, *args)  # type:ignore[arg-type]
+
+    def get_values(self, *args: float | np.ndarray | AdArray) -> np.ndarray:
+        """Fetches the values stored for this secondary operator at its time or iterate
+        index."""
+        return np.hstack([self.fetch_data(self, g, False) for g in self.domains])
+
+    def get_jacobian(self, *args: float | np.ndarray | AdArray) -> sps.spmatrix:
+        """Fetches the derivative values stored for this secondary operator at its time
+        or iterate index.
+
+        Uses the structure of the Jacobians of arguments to insert the values,
+        assuming the Jacobians contain only identity blocks.
+
+        """
+
+        derivatives = np.hstack([self.fetch_data(self, g, True) for g in self.domains])
+
+        # list of jacs per dependency, assuming porepy.ad makes consistent shapes
+        jacs: list[sps.csr_matrix] = []
+
+        for arg, d_i in zip(args, derivatives):
+            if isinstance(arg, AdArray):
+                # At this point we assume that the derivative has only an identity block
+                # Shape checks for derivatives are done in the factory class
+                jacs.append(
+                    sps.csr_matrix((d_i, arg.jac.nonzero()), shape=arg.jac.shape)
+                )
+
+        return sum(jacs).tocsr()
+
+
+class SecondaryExpression:
```

Comment:
Is it correct that this is assumed used as a mixin for a model class, where (due to the `__call__` method) will function similar to any standard method defining a constitutive expression? At least the part about the mixin should be explicitly documented (or have I missed it?), as this seems to be an implicit assumption on usage.